lib/types.ts

```diff
@@ -43,52 +50,55 @@ export type EnvSchemaPropertyValue<
  * Multiple phases may produce errors, then it's stored as array.
  */
 export type EnvSchemaErrors<S extends BaseEnvSchema> = Partial<{
-  [K in keyof S['properties'] | '$other']: Error[];
+  [K in KeyOf<S['properties']> | '$other']: Error[];
 }>;
 export type EnvSchemaMaybeErrors<S extends BaseEnvSchema> =
   | EnvSchemaErrors<S>
   | undefined;
 
 /**
- * Parse one property from string to the best JSONSchema7Type.
+ * Parse one property from string to the best ReadonlyJSONSchema.
  *
  * There is no need to coerce types as Ajv.validate() will do that
  * for you.
  */
 export type EnvSchemaParserFn<
   S extends BaseEnvSchema,
-  K extends keyof S['properties'],
+  V extends BaseEnvParsed<S> = TypeFromJSONSchema<S>,
```

Comment:
here for example, if we always use `TypeFromJSONSchema<S>` anyway, we could simply
```ts
export type EnvSchemaParserFn<
  S extends BaseEnvSchema,
> = (
  str: string,
  propertySchema: Readonly<S['properties'][KeyOf<S['properties']>]>,
  key: KeyOf<S['properties']>,
  schema: Readonly<S>,
) => EnvSchemaPropertyValue<S, TypeFromJSONSchema<S>>;
```

But if you do this on `EnvSchemaPropertyValue`, you won't even need to do this